src/compiler/builder.ts

```diff
@@ -0,0 +1,528 @@
+                    if (referencesInFile.has(removedFilePath)) {
+                        // add files referencing the removedFilePath, as changed files too
+                        const referencedByInfo = fileInfos.get(filePath);
+                        if (referencedByInfo) {
+                            registerChangedFile(filePath as Path, referencedByInfo.fileName);
+                        }
+                    }
+                });
+                // Delete the entry for the removed file path
+                references.delete(removedFilePath);
+            }
+
+            function getReferencedByPaths(referencedFilePath: Path) {
+                return mapDefinedIter(references.entries(), ([filePath, referencesInFile]) =>
+                    referencesInFile.has(referencedFilePath) ? filePath as Path : undefined
+                );
+            }
+
+            function getFilesAffectedByUpdatedShape(program: Program, sourceFile: SourceFile, singleFileResult: string[]): string[] {
+                if (!isExternalModule(sourceFile) && !containsOnlyAmbientModules(sourceFile)) {
+                    return getAllEmittableFiles(program);
+                }
+
+                const options = program.getCompilerOptions();
+                if (options && (options.isolatedModules || options.out || options.outFile)) {
+                    return singleFileResult;
+                }
+
+                // Now we need to if each file in the referencedBy list has a shape change as well.
+                // Because if so, its own referencedBy files need to be saved as well to make the
+                // emitting result consistent with files on disk.
+
+                const seenFileNamesMap = createMap<string>();
+                const setSeenFileName = (path: Path, sourceFile: SourceFile) => {
+                    seenFileNamesMap.set(path, sourceFile && shouldEmitFile(sourceFile) ? sourceFile.fileName : undefined);
+                };
+
+                // Start with the paths this file was referenced by
+                const path = sourceFile.path;
+                setSeenFileName(path, sourceFile);
+                const queue = getReferencedByPaths(path);
+                while (queue.length > 0) {
+                    const currentPath = queue.pop();
+                    if (!seenFileNamesMap.has(currentPath)) {
+                        const currentSourceFile = program.getSourceFileByPath(currentPath);
+                        if (currentSourceFile && updateShapeSignature(program, currentSourceFile, fileInfos.get(currentPath))) {
+                            queue.push(...getReferencedByPaths(currentPath));
+                        }
+                        setSeenFileName(currentPath, currentSourceFile);
+                    }
+                }
+
+                // Return array of values that needs emit
+                return arrayFrom(seenFileNamesMap.values());

```

Comment:
Should this filter out just the non-`undefined` values? `seenFileNamesMap` will have an `undefined` entry for anything that shouldn't be emitted.